File: 1372-longest-zigzag-path-in-a-binary-tree/1372-longest-zigzag-path-in-a-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int res = 0;
    int longestZigZag(TreeNode* root) {
        // do dfs 
        // pass the dir as parameter 
        // return only lst or rst based on dir 
        // int l = preorder(root->left,'l');
        // int r = preorder(root->right,'r');
        // res=max({res,l,r});
        preorder(root,'-');
        return res;
    }
    
    int preorder(TreeNode* node,char dir){
        if(!node) return 0;
        int l = preorder(node->left,'l');
        int r = preorder(node->right,'r');
        res=max({res,l,r});
        if(dir=='l') return 1+r;
        return 1+l;
    }

};

File: 1372-longest-zigzag-path-in-a-binary-tree/1372-longest-zigzag-path-in-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    res = 0
    def longestZigZag(self, root: Optional[TreeNode]) -> int:
        self.postorder(root,'-')
        return self.res

    def postorder(self,root,dir)->int:
        if root is None: 
            return 0 
        l = self.postorder(root.left,'l')
        r = self.postorder(root.right,'r')
        self.res=max(l,r,self.res)
        if dir=='l':
            return 1+r 
        return 1+l